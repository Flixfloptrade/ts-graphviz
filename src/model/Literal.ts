import { DotBase } from '../common';
import { escape, quote } from '../utils/dot-rendering';

/**
 * @category Literal
 */
export class Literal extends DotBase {
  public readonly value: string;

  private isHTMLLike: boolean;
  private isNotString: boolean;
  private isQuoteRequired: boolean;
  constructor(value: string | number | boolean) {
    super();
    this.isNotString = typeof value !== 'string';
    let stringValue: string = typeof value === 'string' ? value : value.toString();
    if (this.isNotString) {
      this.isHTMLLike = false;
      this.isQuoteRequired = false;
    } else {
      const trimmed = stringValue.trim();
      this.isHTMLLike = /^<.+>$/ms.test(trimmed);
      if (this.isHTMLLike) {
        stringValue = trimmed;
        this.isQuoteRequired = false;
      } else {
        this.isQuoteRequired = /[#\s":;=\-']/g.test(stringValue);
      }
    }
    this.value = stringValue;
  }

  public toDot(): string {
    if (this.isNotString || this.isHTMLLike) {
      return this.value;
    }
    let value = this.value;
    if (this.isQuoteRequired) {
      value = quote(escape(value));
    }
    return value;
  }
}
