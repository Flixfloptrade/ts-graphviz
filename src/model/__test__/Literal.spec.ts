import { Literal } from '../Literal';

const DoubleQuotedValuePattern = /^".+"$/ms;
const HTMLLikeValuePattern = /^<.+>$/ms;

describe('valueType', () => {
  describe('Literal', () => {
    describe('plain string', () => {
      test.each([
        ['te"st'],
        [
          `
          test
          `,
        ],
      ])('toDot result should be quoted', input => {
        const value = new Literal(input);
        const dot = value.toDot();
        expect(dot).toMatch(DoubleQuotedValuePattern);
        expect(dot).not.toMatch(HTMLLikeValuePattern);
        expect(dot).toMatchSnapshot();
      });
    });

    describe('HTML Like string', () => {
      test.each([
        ['<<B>hoge</B>>'],
        [
          `<
          <U>hoge</U>
          >`,
        ],
        [
          `
          <<I>hoge</I>
          >
          `,
        ],
      ])('toDot result should not be quoted', input => {
        const value = new Literal(input);
        const dot = value.toDot();
        expect(dot).not.toMatch(DoubleQuotedValuePattern);
        expect(dot).toMatch(HTMLLikeValuePattern);
      });
    });
  });
});
