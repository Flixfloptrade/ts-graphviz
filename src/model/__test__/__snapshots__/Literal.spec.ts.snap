// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`valueType Literal plain string toDot result should be quoted 1`] = `"\\"te\\\\\\"st\\""`;

exports[`valueType Literal plain string toDot result should be quoted 2`] = `"\\"\\\\n          test\\\\n          \\""`;
